Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,775 @@
+
+@[simp]
+theorem smul_fst : (c • x).fst = c • x.fst :=
+  rfl
+
+@[simp]
+theorem smul_snd : (c • x).snd = c • x.snd :=
+  rfl
+
+end algebra
+
+/-! Note that the unapplied versions of these lemmas are deliberately omitted, as they break
+the use of the type synonym. -/
+
+section equiv
+
+variable {p α β}
+
+@[simp]
+theorem equiv_fst (x : WithLp p (α × β)) : (WithLp.equiv p (α × β) x).fst = x.fst :=
+  rfl
+
+@[simp]
+theorem equiv_snd (x : WithLp p (α × β)) : (WithLp.equiv p (α × β) x).snd = x.snd :=
+  rfl
+
+@[simp]
+theorem equiv_symm_fst (x : α × β) : ((WithLp.equiv p (α × β)).symm x).fst = x.fst :=
+  rfl
+
+@[simp]
+theorem equiv_symm_snd (x : α × β) : ((WithLp.equiv p (α × β)).symm x).snd = x.snd :=
+  rfl
+
+end equiv
+
+section DistNorm
+
+/-!
+### Definition of `edist`, `dist` and `norm` on `ProdLp`
+
+In this section we define the `edist`, `dist` and `norm` functions on `WithLp p (α × β)` without
+assuming `[Fact (1 ≤ p)]` or metric properties of the spaces `α` and `β`. This allows us to provide
+the rewrite lemmas for each of three cases `p = 0`, `p = ∞` and `0 < p.toReal`.
+-/
+
+
+section EDist
+
+variable [EDist α] [EDist β]
+
+open Classical in
+/-- Endowing the space `WithLp p (α × β)` with the `L^p` edistance. We register this instance
+separate from `ProdLp.instPseudoEMetric` since the latter requires the type class hypothesis
+`[Fact (1 ≤ p)]` in order to prove the triangle inequality.
+
+Registering this separately allows for a future emetric-like structure on `WithLp p (α × β)` for
+`p < 1` satisfying a relaxed triangle inequality. The terminology for this varies throughout the
+literature, but it is sometimes called a *quasi-metric* or *semi-metric*. -/
+instance instProdEDist : EDist (WithLp p (α × β)) where
+  edist f g :=
+    if _hp : p = 0 then
+      (if f.fst = g.fst then 0 else 1) + (if f.snd = g.snd then 0 else 1)
+    else if p = ∞ then
+      edist f.fst g.fst ⊔ edist f.snd g.snd
+    else
+      (edist f.fst g.fst ^ p.toReal + edist f.snd g.snd ^ p.toReal) ^ (1 / p.toReal)
+
+variable {p α β}
+variable (x y : WithLp p (α × β)) (x' : α × β)
+
+@[simp]
+theorem prod_edist_eq_card [DecidableEq α] [DecidableEq β] (f g : WithLp 0 (α × β)) :
+    edist f g = (if f.fst = g.fst then 0 else 1) + (if f.snd = g.snd then 0 else 1) := by
+  convert if_pos rfl
+
+theorem prod_edist_eq_add (hp : 0 < p.toReal) (f g : WithLp p (α × β)) :
+    edist f g = (edist f.fst g.fst ^ p.toReal + edist f.snd g.snd ^ p.toReal) ^ (1 / p.toReal) :=
+  let hp' := ENNReal.toReal_pos_iff.mp hp
+  (if_neg hp'.1.ne').trans (if_neg hp'.2.ne)
+
+theorem prod_edist_eq_sup (f g : WithLp ∞ (α × β)) :
+    edist f g = edist f.fst g.fst ⊔ edist f.snd g.snd := by
+  dsimp [edist]
+  exact if_neg ENNReal.top_ne_zero
+
+end EDist
+
+section EDistProp
+
+variable {α β}
+variable [PseudoEMetricSpace α] [PseudoEMetricSpace β]
+
+/-- This holds independent of `p` and does not require `[Fact (1 ≤ p)]`. We keep it separate
```

Comment:
Instead of just commenting on the content of the lemma, could you first explain what the lemma says?